### ISSUE TITLE
`ClassVar` and `Final` should be forbidden in unions

### ISSUE THREAD

**Issue by Azureblade3808** [microsoft/pyright#5659]
**Describe the bug**
See sample below.

**Code or Screenshots**
```python
from typing import ClassVar, Final

def int_or_none() -> int | None:
    ...

class C:
    a: ClassVar[int] | None = int_or_none()  # should be invalid, may be a typo of `ClassVar[int | None]`
    b: Final[int] | None = int_or_none()  # should be invalid, may be a typo of `Final[int | None]`
    
    def __init__(self, /) -> None:
        self.a = 0  # should be reported if `a` is class-var
        self.b = 0  # should be reported if `b` is final
```

**VS Code extension or command-line**
Pylance v2023.8.11 Pre-Release
Pyright 1.1.320

**Comment by erictraut** [comment]
There are boundless ways to use `ClassVar` and `Final` and other non-class annotations in ways that have no valid meaning. Detecting and emitting errors in all such cases isn't really viable. Pyright already checks for the most obvious misuses and flags them as errors. I'm not inclined to add more special-purpose checks unless there's evidence that it's a common source of misunderstanding or misuse. I don't think this meets the bar, so I'm going to close this issue. I'm willing to reconsider it in the future based on evidence to the contrary.
